util/src/main/scala/Money.scala

```diff
@@ -576,25 +570,29 @@ object HighPrecisionMoney {
   def factor(fractionDigits: Int): BigDecimal = Money.cachedCentFactor(fractionDigits)
   def centFactor(currency: Currency): BigDecimal = factor(currency.getDefaultFractionDigits)
 
+  private def amountToPreciseAmount(amount: BigDecimal, fractionDigits: Int): Long =
+    (amount * Money.cachedCentPower(fractionDigits)).toLongExact
+
   def fromDecimalAmount(amount: BigDecimal, fractionDigits: Int, currency: Currency)(implicit
       mode: RoundingMode): HighPrecisionMoney = {
     val scaledAmount = amount.setScale(fractionDigits, mode)
-
-    HighPrecisionMoney(scaledAmount, fractionDigits, roundToCents(scaledAmount, currency), currency)
+    val preciseAmount = amountToPreciseAmount(scaledAmount, fractionDigits)
+    val newCentAmount = roundToCents(scaledAmount, currency)
+    HighPrecisionMoney(preciseAmount, fractionDigits, newCentAmount, currency)
   }
 
+  private def centToPreciseAmount(centAmount: Long, fractionDigits: Int, currency: Currency): Long =
+    Math.pow(10, fractionDigits - currency.getDefaultFractionDigits).toLong * centAmount
```

Comment:
Math.pow uses and returns Double types, but it seems to work well for this case until it hits Long.MaxValue
```
(0 to 20).map(i => (i, Math.pow(10, i).toLong))
val res16: IndexedSeq[(Int, Long)] = Vector((0,1), (1,10), (2,100), (3,1000), (4,10000), (5,100000), (6,1000000), (7,10000000), (8,100000000), (9,1000000000), (10,10000000000), (11,100000000000), (12,1000000000000), (13,10000000000000), (14,100000000000000), (15,1000000000000000), (16,10000000000000000), (17,100000000000000000), (18,1000000000000000000), (19,9223372036854775807), (20,9223372036854775807))
```

Reply:
Would it make sense to recover somehow whenever we know it is to power 19+?

Reply:
Or at least add a specific error and a spec for it could also be a case, just so we know it is expected and not by accident.

Reply:
Yes, that's probably a good idea, even just throwing an exception is better than getting weird results.